### PR TITLE
Rewrite the PullSerializer to fix nasty concurrency bugs

### PR DESCRIPTION
**Pull request by pleroy**
Also much better testing.

### CODE REVIEW

File: base/bytes.hpp [hunk 1]
```diff
@@ -10,16 +10,10 @@ namespace base {
...
-  // A |Bytes| object of size 0.
-  static Bytes const Null;
-
- private:
-  // Placeholder for an empty |Bytes| object.
-  static std::uint8_t null_data_;
+  Bytes();  // An object of size 0.
+  Bytes(not_null<std::uint8_t*> const data, int const size);
+  not_null<std::uint8_t*> data;
+  int size;
```
Comment (eggrobin) [inline base/bytes.hpp]: try a const size.

File: base/pull_serializer.hpp [hunk 2]
```diff
@@ -54,16 +49,17 @@ class DelegatingTwoArrayOutputStream
...
 class PullSerializer {
  public:
-  // The |size| of the data objects returned by |Get| are never greater than
-  // |max_size|.  This class uses at most |2 * max_size| bytes (plus some small
-  // overhead).
-  explicit PullSerializer(int const max_size);
+  // The |size| of the data objects returned by |Pull| are never greater than
+  // |chunk_size|.  At most |number_of_chunks| chunks are held in the internal
+  // queue.  This class uses at most
+  // |number_of_chunks + (chunk_size + O(1)) + O(1)| bytes.
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: wat

File: base/pull_serializer_body.hpp [hunk 3]
```diff
@@ -13,35 +13,32 @@ namespace base {
...
+inline bool DelegatingArrayOutputStream::Next(void** const data,
+                                              int* const size) {
+  if (position_ == bytes_.size) {
     // We're at the end of the array.  Hand the current array over to the
-    // callback and start filling the other array.
-    on_full_(Bytes(data1_, size_));
+    // callback and start filling the next one.
+    bytes_ = on_full_(bytes_);
     position_ = 0;
     last_returned_size_ = 0;
```
Comment (eggrobin) [inline base/pull_serializer_body.hpp]: ?

File: base/pull_serializer_body.hpp [hunk 4]
```diff
@@ -52,70 +49,92 @@ void DelegatingTwoArrayOutputStream::BackUp(int count) {
...
+    bytes_ = on_full_(Bytes(bytes_.data, position_));
     position_ = 0;
   }
   last_returned_size_ = 0;
-  swap(data1_, data2_);
 }
 
-std::int64_t DelegatingTwoArrayOutputStream::ByteCount() const {
+inline std::int64_t DelegatingArrayOutputStream::ByteCount() const {
   return position_;
```
Comment (eggrobin) [inline base/pull_serializer_body.hpp]: is this correct?

File: base/bytes.hpp (hunk 1 above)
Comment (eggrobin) [inline base/bytes.hpp]: also int64_t.

File: base/pull_serializer_body.hpp [hunk 5]
```diff
@@ -52,70 +49,92 @@ void DelegatingTwoArrayOutputStream::BackUp(int count) {
...
+inline PullSerializer::PullSerializer(int const chunk_size,
+                                      int const number_of_chunks)
+    : chunk_size_(chunk_size),
+      number_of_chunks_(number_of_chunks),
+      data_(std::make_unique<std::uint8_t[]>(chunk_size_ * number_of_chunks_)),
+      stream_(Bytes(data_.get(), chunk_size_),
               std::bind(&PullSerializer::Push, this, _1)),
-      done_(false) {}
+      is_first_pull_(true) {
+  // Mark all the chunks as free.  The 0th chunk has been passed to the strem,
```
Comment (eggrobin) [inline base/pull_serializer_body.hpp]: typo

File: base/pull_serializer.hpp [hunk 6]
```diff
@@ -72,28 +68,47 @@ class PullSerializer {
...
+
+  // The |queue_| contains the |Bytes| objects filled by |Push| and not yet
+  // consumed by |Pull|.  If a |Bytes| object has been handed over to the caller
+  // by |Pull| it stays in the queue until the next call to |Pull|, to make sure
+  // that the pointer is not reused while the caller processes it.
+  std::queue<Bytes> queue_ GUARDED_BY(lock_);
+
+  // The |free_| queue contains the start addresses of chunks that are not yet
+  // ready to be returned by |Pull|.  That includes the chunk currently being
+  // fille by the stream.
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: typo

File: base/pull_serializer_body.hpp [hunk 7]
```diff
@@ -52,70 +49,92 @@ void DelegatingTwoArrayOutputStream::BackUp(int count) {
...
     std::unique_lock<std::mutex> l(lock_);
-    holder_is_full_.wait(l, [this]() { return done_ || holder_ != nullptr; });
-    if (holder_ != nullptr) {
-      result = std::move(holder_);
+    // The element at the front of the queue is the one that was last returned
+    // by |Pull| and must be dropped and freed, except the first time |Pull| is
+    // called.
+    if (is_first_pull_) {
+      queue_has_elements_.wait(l, [this]() { return !queue_.empty(); });
+      is_first_pull_ = false;
```
Comment (eggrobin) [inline base/pull_serializer_body.hpp]: sentinel?

File: base/pull_serializer.hpp (hunk 2 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/pull_serializer_body.hpp (hunk 3 above)
Comment (pleroy) [inline base/pull_serializer_body.hpp]: Removed.

File: base/pull_serializer_body.hpp (hunk 5 above)
Comment (pleroy) [inline base/pull_serializer_body.hpp]: Done.

File: base/pull_serializer.hpp (hunk 6 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/bytes.hpp (hunk 1 above)
Comment (pleroy) [inline base/bytes.hpp]: #471 for constness.  int64_t done.

File: base/pull_serializer_body.hpp (hunk 4 above)
Comment (pleroy) [inline base/pull_serializer_body.hpp]: Fixed.

**Pull request merged**